### ISSUE TITLE
problem with backticks

### ISSUE THREAD

**Issue by prosserg** [jleyva/moodle-block_configurablereports#117]
we make use of custom views eg

create view mdl_user_data as select `u`.`address` AS 'Hub Access',concat((select `auid`.`data` from (`mdl_user_info_data` `auid` join `mdl_user_info_field` `auif` on(`auid`.`fieldid` = `auif`.`id`)) where `auif`.`shortname` = 'CA Student number' and `auid`.`userid` = `u`.`id`)) AS 'Banner ID' from `mdl_user` `u` where `u`.`deleted` = 0

[while the code above was entered with all tables and fields quoted with backticks - they get removed in the display in this comment box, which is frustrating ]

an sql CR like this

select * from prefix_user_data ud where ud.`hub access` <> "" [again backticks around hub access not displayed here ]

cannot be viewed in ver 2019021500 (with moodle 3.6)  but can be viewed in ver 2016020103 (with moodle 3.1.6)

NB the two moodles are on the same server having Mariadb 10.3.14 PHP 7.0

in 2019021500 the CR can be saved/created but NOT viewed

on view report

Debug info: You have an error in your SQL syntax;
select * from mdl_v2_user_data ud where ud. hub access <> "" LIMIT 0, 5000

NOTE - the backticks have been removed in the CR error report

If I run that same query from a mysql commandline the query works fine

**Comment by prosserg** [comment]
So I found the reason for the change in behaviour between ver 2016020103 and ver 2019021500

in configurable_reports/locallib.php
ver 2016020103 has

    220 function cr_unserialize($var) {
    221     return urldecode_recursive(unserialize($var));
    222 }

while 
ver 2019021500 has

    165 function cr_unserialize($var) {
    166     // It's needed to convert the object to stdClass to avoid __PHP_Incomplete_Class error.
    167     $var = preg_replace('/O:6:"object"/', 'O:8:"stdClass"', $var);
    168     // To make SQL queries compatible with PostgreSQL it's needed to replace " to '.
    169     $var = preg_replace('/THEN\+%22(.+?)%22/', 'THEN+%27${1}%27', $var);
    170     $var = preg_replace('/%60/', '+++', $var);
    171
    172     return urldecode_recursive(unserialize($var));
    173 }

The change is apparantly for "168     // To make SQL queries compatible with PostgreSQL it's needed to replace " to '." although not entirely clear if 170 applies

maybe the code could be

**if($CFG->dbtype    == 'pgsql'){**
    169     $var = preg_replace('/THEN\+%22(.+?)%22/', 'THEN+%27${1}%27', $var);
    170     $var = preg_replace('/%60/', '+++', $var);
**}**
    171

**Comment by sarjona** [comment]
Hi @prosserg!

I was trying to avoid conditionals like the one you proposed based on the DB type because they make harder to maintain the code :-(
I've not been able to reproduce the error with a default Moodle installation to try to find a solution. Could you please provide detailed testing instructions? Maybe you can attach some file to avoid the backticks to disappear also here ;-)
Thanks in advance!!

**Comment by prosserg** [comment]
Hi Sarjona

Here are some files and screen grabs

view_sql.txt is the sql of my view

[view_sql.txt](https://github.com/jleyva/moodle-block_configurablereports/files/3247503/view_sql.txt)
 
2019-06-03 at 13 11 52 is a screen grab of a simple CR calling that view where the configurable_reports/version.php has $plugin->version = 2016020103;

![Screenshot 2019-06-03 at 13 11 52](https://user-images.githubusercontent.com/36161064/58801302-085ed500-8602-11e9-8d32-49b8a9308193.png)

2019-06-03 13 12 13 is a screen grab of 'View Report' 

![Screenshot 2019-06-03 at 13 12 13](https://user-images.githubusercontent.com/36161064/58801359-3ba16400-8602-11e9-97b0-9244d5bf8cd2.png)

IF I then create the same view and same CR in a moodle 3.6 where configurable_reports/version.php has $plugin->version = 2019021500; the CR can be saved but not viewed

2019-06-03 13 31 37 is a screen grab of 'View Report'

![Screenshot 2019-06-03 at 13 31 37](https://user-images.githubusercontent.com/36161064/58802008-fa11b880-8603-11e9-8d16-07d15809f47e.png)

 mysql_log.txt contains a snippet from the mariadb server log captured as the 'View Report' button was pressed.

[mysql_log.txt](https://github.com/jleyva/moodle-block_configurablereports/files/3247595/mysql_log.txt)

 2019-06-03 13 51 09  shows the result of running the same CR sql (but with backticks preserved) directly in the mysql console

![Screenshot 2019-06-03 at 13 51 09](https://user-images.githubusercontent.com/36161064/58803174-b79dab00-8606-11e9-812c-075e4c51566f.png)


Does this information help ?

**Comment by sarjona** [comment]
Thanks for the information! 

Which is the plugin installing the prefix_v3_user_data tables? It's weird because database table column names cannot contain a space, like the "hub access" displayed in the screenshots.

**Comment by prosserg** [comment]
Thanks for the quick response!

I make the mdl_v3_user_data VIEW using the sql in the file I attached (see above) view_sql.txt from within phpmyadmin.

If I change that sql so instead of `Hub Access` I have `Hub_Access` and if I also change my CR sql from 
select * from prefix_v3_user_data ud where ud.`hub access`       <> ""
to
select * from prefix_v3_user_data ud where ud.`hub_access`       <> ""

then the 'View Report' error goes away. Fine.

But if I THEN look at the CR 'Custom SQL' tab, the sql code has been changed and looks like this

![Screenshot 2019-06-03 at 14 30 01](https://user-images.githubusercontent.com/36161064/58805555-347f5380-860c-11e9-84f5-c9be4bd0cb67.png)

In other words the configurable reports code is removing the backticks

Is that intended ?
Gary

**Comment by prosserg** [comment]
Sorry once again this html editor is ALSO removing backticks !

The CR sql change is in this file

[CR_sql_change.txt](https://github.com/jleyva/moodle-block_configurablereports/files/3247803/CR_sql_change.txt)

**Comment by sarjona** [comment]
Hi Gary!
Thanks a lot for the explanation (and my apologies for the late reply). If I've understood it properly, I would say that this is the expected behaviour: as MySQL is also working without the backticks, they shouldn't be used.

Is OK for you if I close the issue or is there anything more to do here? :-) 

Kind regards from Barcelona,
Sara

**Comment by prosserg** [comment]
Hi Sara

I am coming back to this issue because we still have problems. Our main problem is migrating our 25 moodle sites to latest configurable_reports code given that we have our own code in a bunch of CR sql reports calling views.

You say that mysql / mariadb "should not use backticks" - yet the latest version of both DOES allow backticks.

The php code since ver 2019021500 of the CR plugin explicitly removes backticks from custom sql CR sql. This is apparantly solely because pgsql (unlike mysql / mariadb) does not allow backticks.

Is the use case of pgsql backend larger than mysql/mariadb ? Because your choice of 

165 function cr_unserialize($var) {
166     // It's needed to convert the object to stdClass to avoid __PHP_Incomplete_Class error.
167     $var = preg_replace('/O:6:"object"/', 'O:8:"stdClass"', $var);
168     **// To make SQL queries compatible with PostgreSQL it's needed to replace " to '.**
169     $var = preg_replace('/THEN\+%22(.+?)%22/', 'THEN+%27${1}%27', $var);
170     $var = preg_replace('/%60/', '+++', $var);
171
172     return urldecode_recursive(unserialize($var));
173 }

lines 169,170 is *purely* for pgsql and that backend db is possibly a minority of use cases.

When I asked if you could amend the code to

if($CFG->dbtype == 'pgsql'){
169 $var = preg_replace('/THEN+%22(.+?)%22/', 'THEN+%27${1}%27', $var);
170 $var = preg_replace('/%60/', '+++', $var);
}

you said

"I was trying to avoid conditionals like the one you proposed based on the DB type because they make harder to maintain the code :-("

Meanwhile you have imposed a problem for mysql/mariadb users - the problem we have is that our views no longer work. 

BUT since standard moodle code (which does not have any column names requiring backticks) I don't get what is the **necessity** for your new version of function cr_unserialize. Your code penalises legacy users with mysql/mariadb (perhaps even mssql) while preferring pgsql users.

The wonderful aspect of the configurable_reports plugin is that is allows creativity via the customsql report type. That is really appreciated. But the 'blanket' removal of backticks from report code, solely for pgsql backends, is I think, a mistake.

Please, please, could you reconsider.

Very best wishes
Gary

**Comment by prosserg** [comment]
To summarise my previous post

The latest code in the wonderful configurable_reports plugin penalises users with working custom code and mysql / mariadb backends in favour of one backend - pgsql - using a blanket applied *fix* which for pgsql backends could be easily addressed by a conditional statement.

Application of a code fix applied across all backends, should take account of working code (even if deemed legacy) when the mitigation is a mere 2 lines of code.

Gary
